### PR TITLE
Fix mismatched shape in convolution_backward_overrideable

### PR DESCRIPTION
**Pull request by Krovatkin**
Fixes #{issue number}

### CODE REVIEW

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp [hunk 1]
```diff
@@ -319,9 +319,17 @@ class TSNodeLowering : public TSNodeLoweringInterface {
             ? LowerBuiltin(at::aten::cudnn_convolution_transpose_backward,
                            arguments)
             : LowerBuiltin(at::aten::cudnn_convolution_backward, arguments);
-    // cudnn_convolution_backward/cudnn_convolution_transpose_backward only
-    // returns 2 tensors
-    result.push_back(nullptr);
+
+    // N.B. rank is specialized even with dynamic shapes
+    // all axes but channel to reduce to the bias size
+    auto grad_rank = dynamic_cast<const ir::TsNode*>(operands[0].node)->shape(operands[0].index).dim();
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:325]: better to use GetShapeFromTsNode or GetShapeFromTsValue helpers here, just so we can easily refactor them away or grep them later

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp [hunk 2]
```diff
@@ -319,9 +319,17 @@ class TSNodeLowering : public TSNodeLoweringInterface {
...
+    std::vector<int64_t> axes(grad_rank);
+    std::iota(axes.begin(), axes.end(), 0);
+    TORCH_INTERNAL_ASSERT(grad_rank >= 3, "Convolution outputs should have 3+ dimensions");
+    axes.erase(axes.begin() + 1);
+    auto axes_val = loctx()->graph()->insertConstant(axes);
+    auto bias_grad = loctx()->graph()->insert(at::aten::sum, {loctx()->GetOutputOp(operands[0]), axes_val});
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:331]: I lack the context for why this change is necessary- we're essentially saying that we need to sum away all but the first dimension of the grad, which i guess is channels?  why isn't this isn't done for us by autograd?

Comment (desertfire) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:331]: We may just get the last output from the generic `convolution_backward`?https://github.com/pytorch/pytorch/pull/65219

Comment (Krovatkin) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:331]: automatic reduction only works if shapes are broadcastable. [3] and [1, 3, 128, 128] aren't broadcastable because we pad 1 from the left. [1, 1, 1, 3] and [1, 3, 128, 128] don't match in 3rd and 4th dim. It works for CUDA because we reshape [3] to [1, 3, 1, 1] before we add the output of a conv to it, so we have to manually either reshape or sum. I just did `sum`.

Comment (Krovatkin) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:331]: > We may just get the last output from the generic convolution_backward

That seems to only compute gradients for input and weight? 
```
- func: cudnn_convolution_backward(Tensor self, Tensor grad_output, Tensor weight, int[] padding, int[] stride, int[] dilation, int groups, bool benchmark, bool deterministic, bool allow_tf32, bool[2] output_mask) -> (Tensor, Tensor)
```

Comment (Krovatkin) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:331]: in case you want to reproduce the issue outside of TB, this example should do:

```
dev = 'lazy'
inp = torch.rand(1, 3, 128, 128, requires_grad = True).to(device=dev)
in_channels, out_channels = 3, 32
grad = torch.rand(1, 32, 63, 63).to(device=dev)
conv = torch.nn.Conv2d(in_channels, out_channels, kernel_size=3, stride=2, bias=True)
conv.to(dev)
conv_out = conv(inp)
conv_out.backward(grad)
```

Comment (desertfire) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:331]: > automatic reduction only works if shapes are broadcastable. [3] and [1, 3, 128, 128] aren't broadcastable because we pad 1 from the left. [1, 1, 1, 3] and [1, 3, 128, 128] don't match in 3rd and 4th dim. It works for CUDA because we reshape [3] to [1, 3, 1, 1] before we add the output of a conv to it, so we have to manually either reshape or sum. I just did `sum` for testing. Now I can run alexnet_train.

Thanks for the explanation, now I can understand better. When you say use `sum`, I guess the implication is `sum` does auto broadcasting, but how does it affect the actual return value?

Comment (desertfire) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:331]: > > We may just get the last output from the generic convolution_backward
> 
> That seems to only compute gradients for input and weight?
> 
> ```
> - func: cudnn_convolution_backward(Tensor self, Tensor grad_output, Tensor weight, int[] padding, int[] stride, int[] dilation, int groups, bool benchmark, bool deterministic, bool allow_tf32, bool[2] output_mask) -> (Tensor, Tensor)
> ```

Yes, that's my understanding too.

**Comment by desertfire** [comment]
Another general comment, can we add a test case in test_ptltc?

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp (hunk 2 above)
Comment (Krovatkin) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:331]: @desertfire 
> guess the implication is sum does auto broadcasting

It does the check that `fwd_op` input's shape can be broadcasted to output's shape (which is also gradient's shape). Then it actually runs automatic `sum` (which is a bwd operation to broadcast) to reduce the gradient into the shape of the input. In our case, the input shape is [3] and the gradient shape [1, 3, 128, 128]. So `is_expandable` check in
https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L682-L689 fails.

Since we can't rely on the automatic reduction (`sum`) we either have to do it manually which I did in this fix. This is frankly the easiest because the fix is very local to the backward function.

I mentioned we could probably try imitating what CUDA does (e.g. a manual reshape of [3] to [1, 3, 1, 1] but I'm not 100% sure how to pull it off, since by the time we lower to TS we already built the backward graph and for convolution_overrideable we would only run `convolution_overrideable_backward` and that takes [3] bias. I don't know how to "insert" `view` that would be visible to the autograd if it makes sense.

Comment (Krovatkin) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:331]: > Yes, that's my understanding too.

Yes, so AFAIU we need to handle bias ourselves (outside of `cudnn_convolution_backward`). thankfully, the bwd of addition is just the gradient itself. The only mishap here is with shapes. 😢

**Review by desertfire — APPROVED** [review]

File: lazy_tensor_core/test/test_lazy.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                        return copy_t
+
+                inp = torch.rand(1, 3, 128, 128, device='cuda', requires_grad = True)
+                inp_copy = clone_move(inp)
+                grad = torch.rand(1, 32, 121, 121, device='cuda') # no requires_grad
+                grad_copy = clone_move(grad)
+                weight = torch.rand(32, 3, 8, 8, device='cuda', requires_grad = True)
+                weight_copy = clone_move(weight)
+                bias = torch.rand(32, device='cuda', requires_grad = True)
+                bias = bias
```
Comment (wconstab) [inline lazy_tensor_core/test/test_lazy.py:28]: what's this?

File: lazy_tensor_core/test/test_lazy.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+                # run eager
+                conv_out = torch.nn.functional.conv2d(inp, weight, bias)
+                (inp_grad, weight_grad, bias_grad) = torch.autograd.grad([conv_out], [inp, weight, bias], [grad])
+
+                # run lazy
+                conv_copy_out = torch.nn.functional.conv2d(inp_copy, weight_copy, bias_copy) 
+                (inp_copy_grad, weight_copy_grad, bias_copy_grad) = torch.autograd.grad([conv_copy_out], [inp_copy, weight_copy, bias_copy], [grad_copy])
+
+                jit_graph = lazy_tensor_core._LAZYC._get_ltc_tensors_backend([bias_copy_grad])
+                FileCheck().check("aten::sum(").run(jit_graph)
```
Comment (wconstab) [inline lazy_tensor_core/test/test_lazy.py:41]: you're checking to make sure the full graph is captured? and if the aten::sum is present, you assume that came from autograd and it means you captured everything?

**Review by wconstab — APPROVED** [review]

File: lazy_tensor_core/test/test_lazy.py (hunk 4 above)
Comment (Krovatkin) [reply lazy_tensor_core/test/test_lazy.py:41]: I was mostly checking that we captured the backward for the bias. There should be only one `sum` in the whole backward graph since there are no other broadcasts but the one on the bias.

File: lazy_tensor_core/test/test_lazy.py (hunk 3 above)
Comment (Krovatkin) [reply lazy_tensor_core/test/test_lazy.py:28]: my bad, will remove

**Review by desertfire — APPROVED** [review]
